### ISSUE TITLE
Spyder was unable to retrieve this varibale from the console and I find a way to solve this

### ISSUE THREAD

**Issue by river7816** [spyder-ide/spyder#20671]
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [x] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [x] Reset preferences with ``spyder --reset``
    * [x] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
<img width="500" alt="image" src="https://user-images.githubusercontent.com/114122443/224472029-ecaad5d7-d32e-4481-9f6c-c819f48206f3.png">
Although you said there is nothing to do about it, I still insist to report it since I think there is a way to fix it. I encounter this problem, I met this problem the same as [https://github.com/spyder-ide/spyder/issues/15998#issue-940766720](url). I encounter this problem in the self created environment if I do not install the package in the base enviroment. For example, if you do not have pandas in your base environment, and you install pandas in your self-created environment, then you create a Dataframe object in your self-created environment. When you try to open the Dataframe, you will encounter this problem.

### What steps reproduce the problem?

1. install a new miniforge in your computer, which must be new, without other package installed, like numpy pandas. 
2. install spyder in your base environment with `conda install spyder=5.4.2`
3. create a new environment with `conda create -n test python=3.10 `
4. install pandas in the test environment with `conda activate test` `conda install pandas`
5. open spyder and switch to test environment
6. create a empty python file and write `import pandas as pd` `a = pd.DataFrame()`, you will find you cannot open it.

### A way to solve this

1. install pandas in the base environment with `conda install pandas`
2. create a empty python file and write `import pandas as pd` `a = pd.DataFrame()` in the base enviroment, you will find you can open it now, but still unnormal. it looks like this picture below:
<img width="955" alt="image" src="https://user-images.githubusercontent.com/114122443/224472901-644ecf21-4347-476f-8351-dda809fbe029.png">
3. back to your test environment, run the same code (you may need to reopen spyder), you can open variabel a now

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: 5.4.2
* Python version: 3.10.9
* Qt version: 5.15.8
* PyQt version: 2.3.0
* Operating System name/version: windows 11 workstation

### Dependencies

```
PASTE DEPENDENCIES HERE
```

**Comment by dalthviz** [comment]
Hi @river7816 thank you for the feedback! I think the message could be improved maybe pointing to some Spyder docs page describing for example the way to workaround this when using a custom interpreter and other possible causes for the message to appear instead of the text saying that you should not create any issues on GitHub (which can sound kind of strange). What do you think @spyder-ide/core-developers ?

Also, I guess this is a kind of duplicate of https://github.com/spyder-ide/spyder/issues/15998 , right?

**Comment by CAM-Gerlach** [comment]
The text should definitely be improved; not only is it not grammatical/idiomatic but it indeed sounds very strange and unhelpful. Beyond that, yeah we could add a FAQ question and have the dialog link to it. Of course, the above won't actually work if the Spyder runtime environment is not user-modifiable, like with our installers.

This is really a two part issue—one part is telling the user what to do when optional dependencies, like Pandas and NumPy are not installed that enable specialized types of viewers, and the other is providing a better way to serialize and represent arbitrary objects that does not rely on unpickaling, and thus require Spyder's runtime environment to have those packages installed. The former case is addressed by either manually or via the installer installing those optional deps, and the latter case by using some thing like JSONpickle, or similar, which I believe we discussed recently but I can't find an issue for.

> Also, I guess this is a kind of duplicate of https://github.com/spyder-ide/spyder/issues/15998 , right?

Not really I don't think, that one and #18821 are more about how CWD. `PATH`, `PYTHONPATH` and `sys.path` interact with Spyder's ability to import custom user modules in the Variable Explorer that were "installed" via PATH hacks, while this one is about the issue of not being able read data types from third party packages installed in the working env but not the base env.

**Comment by ccordoba12** [comment]
> This is really a two part issue—one part is telling the user what to do when optional dependencies, like Pandas and NumPy are not installed that enable specialized types of viewers, and the other is providing a better way to serialize and represent arbitrary objects that does not rely on unpickaling, and thus require Spyder's runtime environment to have those packages installed

I disagree. This issue should only be about improving the message that tells users Spyder was unable to open a variable. @dalthviz's idea about linking to our docs for a more complete explanation and how to sort out the problem on the user's end is a really good one. However, the reworded message should say clearly that this is *not* a Spyder bug, hence it doesn't need to be reported as such (I don't want to deal with issues about this when we know we can really solve them).

The second part you suggested @CAM-Gerlach is totally different from the above one. It's important, sure, but not something we need to address to fix this issue.

I also assigned you @CAM-Gerlach to solve this issue because you disagree with the current message (first part of this issue) and you suggested to add a FAQ to our docs to expand what users can do about it (second part).

**Comment by dalthviz** [comment]
If there is no issue, I think we should create one for the `providing a better way to serialize and represent arbitrary objects` part

**Comment by ccordoba12** [comment]
Agreed. And please ping Quentin about it.

**Comment by CAM-Gerlach** [comment]
Sorry, I hadn't slept much and was in a rush to leave for something, couldn't find the prior issues regarding each of these that I could have sworn we had, and typed out my message hurriedly without thinking through it carefully as I should have.

> This issue should only be about improving the message that tells users Spyder was unable to open a variable. @dalthviz's idea about linking to our docs for a more complete explanation and how to sort out the problem on the user's end is a really good one.

> The second part you suggested @CAM-Gerlach is totally different from the above one. It's important, sure, but not something we need to address to fix this issue.

Agreed, you're right, particularly since the issue here is about an optional dependency for a specific Variable Explorer view (dataframe viewer) rather than the generic object viewer for serializing custom objects. I was just in a rush and not thinking or communicating clearly, sorry.

> However, the reworded message should say clearly that this is not a Spyder bug, hence it doesn't need to be reported as such (I don't want to deal with issues about this when we know we can really solve them).

In most cases yes, but not necessarily all, and its both more effective and more positive to tell users what _to_ do about it rather than what _not_ to do. To note, in some cases it can be fixed on Spyder's side, though. For example, if the version of Pandas Spyder depends on is too old, it won't be able to open newer Pandas dataframes when upicked, e.g. there's a specific compat issue when trying to unpickle Pandas 1.5.x (or maybe 1.4.x) dataframes with 1.3.x that I ran into myself the other day. There's also some edge cases I've hit in the past where the message showed when the variable explorer got stuck loading another variable and then couldn't display others (which wasn't easy for me to repro, unfortunately).

> I also assigned you @CAM-Gerlach to solve this issue because you disagree with the current message (first part of its solution) and you suggested to add a FAQ to our docs to expand what users can do about it (second part).

Yes, I will add it to my queue to work on soon, both the UI text and creating a docs FAQ for it to link do.

> If there is no issue, I think we should create one for the `providing a better way to serialize and represent arbitrary objects` part

Yeah, I could have sworn there's at least one if not multiple issues discussing this, but despite a lot of searching previously I couldn't find one.

> And please ping Quentin about it.

@impact27

**Comment by ccordoba12** [comment]
> In most cases yes, but not necessarily all, and its both more effective and more positive to tell users what to do about it rather than what not to do. 

True, but we already have to deal with so many issues that I wouldn't like to deal with more due to this change. But I guess it really depends on how the new message is worded. So, we'll see.

> And please ping Quentin about it.

My idea with that message was avoiding to ping Quentin on this issue, which doesn't concern him too much, and reserve that instead for the new issue @dalthviz is going to open. So, for next time, please avoid this kind of corrections because it's a bit bothersome to be pinged on issues you have not much to say.

**Comment by ryantrhoades** [comment]
I find it funny how it's been **years** and y'all still haven't made this error message different or any more informative

**Comment by MrAnimaniac** [comment]
Thank you for this nice issue report. I ran into a comparable problem using [lmfit](https://lmfit.github.io/lmfit-py/). It took me a while to track this package down as being the culprit, as some lmfit elements were nested in a pandas data frame. 
Could the improved message hint towards the package missing in the base environment?

**Comment by ccordoba12** [comment]
I think we should be able to do that, but I'm not really sure at the moment.

**Comment by gaussian11** [comment]
This report is really helpful for me.
I could solve this problem at my conda environment.
How to solve:
I has typed just one line on anaconda prompt
>(base) conda install pandas
It was so simple solution right?
However, what actually makes troubleshooting the most difficult is the error message. 
Why are they saying there is nothing on GitHub?

 
![스크린샷 2024-11-16 231856](https://github.com/user-attachments/assets/017cc424-7bb9-4393-b750-7a0bb3659240)
